helix-core/src/main/java/org/apache/helix/controller/stages/ExternalViewComputeStage.java

```diff
@@ -96,7 +96,7 @@ public void execute(final ClusterEvent event) throws Exception {
 
     // Keep MBeans for existing resources and unregister MBeans for dropped resources
     if (clusterStatusMonitor != null) {
-      clusterStatusMonitor.retainResourceMonitor(monitoringResources);
+      //clusterStatusMonitor.retainResourceMonitor(monitoringResources);
```

Comment:
First off, great work on debugging this and identifying the root cause!

Have a few points to add:

- First, with a guiding question: conceptually, where would be the "correct" place for metric-cleanup logic to be? 
- We want the metric cleanup logic to be in an async stage in order to not block the critical controller pipeline. As such, adding the metric-cleanup logic in ResourceValidation may not be appropriate (this is not an async stage). If you believe that this logic shouldn't belong in ExternalViewComputeStage, then can we think about a different (or a new) async stage where this logic can be appropriately placed? With that said though, if you place it in another async stage, would we still face the same issue of this logic not triggering in tests?
- If you think ExternalViewComputeStage is an appropriate place for this cleanup logic, can we simply add a CurrentState change event to artificially trigger the stage just for the sake of making the test case work? I think this is also a valid way because in real-life scenarios we can rely on the ExtViewComputeStage to eventually get called asynchronously. We can think of this as making the test environment model the real environment more closely.

Reply:
Thanks for the comment!

> where would be the "correct" place for metric-cleanup logic to be?

With limited context and knowledge, I think both `ResourceValidationStage` and `ExternalViewComputeStage` make sense. MetricsMonitors belong to, or at least associated with resources. On the other hand, generally speaking, it's also a "view". 
it's also applicable to be in a standalone stage instead of mixing up with other logic.

> async stage...

Yeah, this is a good point I wasn't aware of. So essentially, we'd want to run it in async way, and make sure all related cluster events trigger such computation. 

> add a CurrentState change event to artificially trigger the stage

This is definitely an option for the fix. But could it introduce more overhead and side effect? And what the difference from adding `externalViewPipeline` to `IdealStateChange` registry? All we need is refresh the resource monitors, but we will be running a few other pipelines.